W-21066693-dataweave-transformation-policy-lds (#992)
mulesoft/docs-gateway: gateway/1.12/modules/ROOT/pages/policies-included-script-evaluation-transformation.adoc

= Script Evaluation Transformation Policy
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images
:keywords: script evaluation, transformation, api gateway, flex gateway, gateway, policy

> :page-aliases: api-manager::script-evaluation-transformation-policy.adoc, policies::policies-included-script-evaluation-transformation.adoc

Reviewer: ```suggestion
```

[width="100%", cols="5,15"]
|===
>s| Policy name | Script Evaluation Transformation

> >s|Summary      | Evaluates a DataWeave script on request or response to allow or block the message.

Reviewer: ```suggestion
>s|Summary      | Evaluates a DataWeave script on request or response to allow or block the message
```

>s|Category | Transformation

> >s| First Flex Gateway version available | 1.12

Reviewer: ```suggestion
>s| First Flex Gateway version available | v1.12.0
```

.1+>.^s| Returned Status Codes
| `400` - The DataWeave script failed to execute or returned an invalid result.
|===

== Summary

The Script Evaluation Transformation policy evaluates a DataWeave script on the request or response, including headers and optionally the body, to allow or block the message. The script returns an object with a Boolean `success` value. If `success` is `false`, the script can return a `response` object with `statusCode`, `body`, and `headers`. The gateway sends that as the immediate HTTP response to the client and stops processing. If `success` is `true`, the message continues.

> Result format. 

Reviewer: ```suggestion
The script returns an object with a Boolean `success` value.  If `success` is `true`, the message continues. If `false`, the script can returns a `response` object with the `statusCode`, `body`, and `headers`. The gateway immediately ends the request and returns the `response` object to the client:

Script result format:

```

* Success: `{ "success": true }`. 
* Error: `{ "success": false, "response": { "body": "...", "headers": {}, "statusCode": 400 } }`.

== Configure Policy Parameters

> The script accesses headers and optionally the body. It returns an object with a Boolean `success` field. If `success` is `false`, the script can include a `response` object with `statusCode`, `body`, and `headers`. If `success` is `true`, the script doesn't return a `response` object.

Reviewer: ```suggestion
```

> In Local Mode, apply the policy via declarative configuration using the `script-evaluation-transformation` policy. In Managed Flex Gateway and Flex Gateway Connected mode, apply the policy from the UI and select the script evaluation type.

Reviewer: ```suggestion
```

> include::partial$policy-title-headers.adoc[tag=configFileTitleOnly]

Reviewer: ```suggestion
include::partial$policy-title-headers.adoc[tag=configFile]
```

----
- policyRef:
    name: script-evaluation-transformation
  config:
    script: <string>           // REQUIRED - DataWeave script
    requestFlow: <string>      // REQUIRED - "onRequest" or "onResponse"
    requiresPayload: <boolean> // OPTIONAL, default: false
----

[%header%autowidth.spread,cols="a,a,a,a"]
|===
|Parameter | Required or Optional | Default Value | Description

| `script`
| Required
| —

> | A DataWeave script that returns an object with a boolean `success` field. If `success` is `false`, the optional `response` is: `{ "body": string, "headers": object, "statusCode": number }`.

Reviewer: ```suggestion
| A DataWeave script that returns an object with a boolean `success` field.
```

| `requestFlow`
| Required
| —

> | The phase when the policy runs: `onRequest` or `onResponse`.

Reviewer: ```suggestion
| Whether the policy transforms the request or response body. Supported values: `onRequest` or `onResponse`.
```

| `requiresPayload`
| Optional
| false

> | If true, the script receives the message body in the `payload` binding. If true and the body is missing, the gateway returns an error.

Reviewer: ```suggestion
| If true, the script receives the message body in the `payload` binding. If true and the body is missing, the gateway returns a `400` error.
```

|===

=== Resource Configuration Example

----
  policies:
    - policyRef:
        name: script-evaluation-transformation
      config:
        script: |
          %dw 2.0
          output application/java
          ---
          if (attributes.headers["x-api-key"] != null) { success: true }
          else { success: false, response: { statusCode: 401, body: "Missing API key", headers: {} } }
        requestFlow: "onRequest"
        requiresPayload: false
----

include::partial$policy-title-headers.adoc[tag=ui]

[%header%autowidth.spread,cols="a,a"]
|===
| Parameter | Description

> | Policy type | Script evaluation transformation

Reviewer: ```suggestion
```

| Script | The DataWeave script that returns a success or error result.

> | Request flow | The flow where the policy applies: request or response.

Reviewer: ```suggestion
| Request flow | Whether the policy transforms the request or response body.
```

| Requires payload | Whether the script requires the message body.

> | Method & Resource conditions | The specific methods and resources to which the policy is restricted.

Reviewer: ```suggestion
```
